### PR TITLE
[BE]: Enable F821 and fix bugs

### LINKED ISSUE #112371: Why is flake8 F821 disabled
Flake8 F821 "undefined name" is one of the most useful flake8 rules (see recent https://github.com/pytorch/pytorch/pull/112336 as an example).

Yet the rule is disabled:
https://github.com/pytorch/pytorch/blob/bbd5b935e49a54578ac88cb23ca962ab896a8c7a/.flake8#L10
https://github.com/pytorch/pytorch/blob/bbd5b935e49a54578ac88cb23ca962ab896a8c7a/pyproject.toml#L43

I think the rule needs to be at least selectively enabled. There are currently multiple actual issues in the codebase that can be caught by this rules.

### PR DESCRIPTION
**Pull request by Skylion007**
Fixes #112371 

I tried to fix as many of the bugs as I could, a few I could not figure out what the proper fix for them was though and so I left them with noqas.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @LucasLLC

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1577,6 +1577,9 @@ def _unimplemented():
                     # not broadcastable, can't be compared
                     _unimplemented()
             tensor_cls = left if isinstance(left, TensorVariable) else right
+            proxy = tx.output.create_proxy(
```
Comment (Skylion007) [inline torch/_dynamo/variables/builtin.py:1580]: Seems our tests never caught this one. Not sure how to make it catch it either.

**Review by ezyang — APPROVED** [review]

**Comment by kit1980** [comment]
Great, thank you for this!

**Comment by kit1980** [comment]
Do we want to eventually remove the `test/**: F821` ignore?

**Pull request closed without merging**